File: pset1/credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>

int digits_in_ll(long long ll);
bool pass_luhn_algo(long long ll);
int two_leading_digits(long long ll);
bool is_cc(long long cc);
bool is_known_cc(int digits, int lead);
void newline(void);

int main(void)
{
    printf("Number: ");
    long long cc = get_long_long();
    
    is_cc(cc);
    
    newline();
}

bool is_cc(long long cc)
{
    bool is_cc = pass_luhn_algo(cc);
    int digit_count = digits_in_ll(cc);
    int lead_digits = two_leading_digits(cc);
    
    //is_cc? printf("Could be a credit card number.") : printf("Is not a credit card number.");
    //newline();
    //printf("Has %i digits, leading two are %i.\n", digit_count, lead_digits);
    
    if(is_known_cc(digit_count, lead_digits))
    {
        return true;
    }
    else
    {
        printf("INVALID");
        return false;
    }
}

bool pass_luhn_algo(long long ll)
{
    int total = 0;
    bool odd = true;
    for (long long i = 1; i <= ll; i *= 10)
    {
        int digit = (ll / i) % 10;

        if (odd)
        {
            total += digit;
        }
        else
        {
            digit = digit * 2;
            if (digit < 10)
            {
                total += digit;
            }
            else
            {
                total += digit % 10;
                total += round(digit / 10);
            }
        }

        odd = !odd;
    }

    return (total % 10) == 0;
}

bool is_known_cc(int digits, int lead)
{
    switch(lead)
    {
        case 34:
        case 37:
            if (digits == 15)
            {
                printf("AMEX");
                return true;
            }
            break;
        
        case 51:
        case 52:
        case 53:
        case 54:
        case 55:
            if (digits == 16)
            {
                printf("MASTERCARD");
                return true;
            }
            break;
        
        case 40:
        case 41:
        case 42:
        case 43:
        case 44:
        case 45:
        case 46:
        case 47:
        case 48:
        case 49:
            if (digits == 13 || digits == 16)
            {
                printf("VISA");
                return true;
            }
            break;
    }
    
    return false;
}

int two_leading_digits(long long ll)
{
    while (ll >= 100)
    {
        ll = round(ll / 10);
    }
    
    return (int) ll;
}

int digits_in_ll(long long ll)
{
    int digits = 0;
    for (long long i = 1; i <= ll; i *= 10)
    {
        digits++;
    }
    return digits;
}

void newline(void)
{
    printf("\n");
}

File: pset6/credit.py
import cs50

def main():
    print('Number: ', end='')
    cc = cs50.get_int()
    
    cc_type = whose(cc)
    luhn_res = luhn(cc)

    if luhn_res != 0:
        print('Thought it was {}'.format(cc_type))
        print('Failed Luhn\'s algorithm with {}'.format(luhn_res))
        cc_type = 'INVALID'
    
    print(cc_type)

# return boolean value for whether the input cc
# (credit card number) passes luhn's algorithm
def luhn(cc):
    my_cc = str(cc)
    total = 0
    
    # traverse credit card number - will use i as iteration # but not index
    for i in range(len(my_cc)):
        # reverse order
        index = len(my_cc) - i - 1

        # pull out digit at index
        my_digit = int(my_cc[index])
        
        # set amount to add to the digit
        add = my_digit
        
        # on every other digit, double the amount, add digits (max 2)
        if i % 2 == 1:
            add *= 2
            add = add % 10 + add // 10
        
        #print('[Luhn] Total: {} Digit: {} Will Add: {}'.format(total, my_digit, add))
        total += add
    
    #print('[Luhn] Grand Total: {}'.format(total))
    return total % 10

# return a string value for the credit card owner
# based on number of digits and leading digits
def whose(cc):
    cc = str(cc)
    ln = len(cc)
    ld = int(cc[0:2])
    fd = int(cc[0:1])
    
    if ln == 15 and ld in [34, 37]:
        return 'AMEX'
    elif ln == 16 and ld in range(51,56):
        return 'MASTERCARD'
    elif ln in [13, 16] and fd == 4:
        return 'VISA'
    
    return 'INVALID'

if __name__ == '__main__':
    main()